File: ui/src/lib/transcription.ts
import { GATEWAY_URL } from './api';
import { applyPreReleaseHeader } from './preRelease';

const WHISPER_ENDPOINT = 'https://chutes-whisper-large-v3.chutes.ai/transcribe';

interface TranscriptionOptions {
  language?: string | null;
}

interface TranscriptionResult {
  text: string;
  language?: string;
  duration?: number;
}

interface TranscriptionErrorDetail {
  error: string;
  code: string;
  recoverable: boolean;
  suggestion?: string;
}

export class TranscriptionFailedError extends Error {
  readonly code: string;
  readonly recoverable: boolean;
  readonly suggestion?: string;

  constructor(
    message: string,
    code: string = 'UNKNOWN',
    recoverable: boolean = true,
    suggestion?: string
  ) {
    super(message);
    this.name = 'TranscriptionFailedError';
    this.code = code;
    this.recoverable = recoverable;
    this.suggestion = suggestion;
  }
}

export async function transcribeAudio(
  audioBlob: Blob,
  options: TranscriptionOptions = {}
): Promise<TranscriptionResult> {
  const base64 = await blobToBase64(audioBlob);

  const apiKey = process.env.NEXT_PUBLIC_CHUTES_API_KEY;
  if (!apiKey) {
    throw new TranscriptionFailedError(
      'Chutes API key not configured',
      'API_KEY_MISSING',
      false,
      'Please type your message instead'
    );
  }

  const payload: Record<string, string> = { audio_b64: base64 };
  if (options.language) {
    payload.language = options.language;
  }

  const response = await fetch(WHISPER_ENDPOINT, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const error = await response.text();
    throw new TranscriptionFailedError(
      `Transcription failed: ${error}`,
      'DIRECT_API_ERROR',
      true,
      'Please try again'
    );
  }

  const result = await response.json();
  return normalizeTranscriptionResult(result);
}

export async function transcribeViaGateway(
  audioBlob: Blob,
  options: TranscriptionOptions = {}
): Promise<TranscriptionResult> {
  const base64 = await blobToBase64(audioBlob);
  const payload: Record<string, string> = { audio_b64: base64 };
  if (options.language) {
    payload.language = options.language;
  }

  const response = await fetch(`${GATEWAY_URL}/api/transcribe`, {
    method: 'POST',
    headers: applyPreReleaseHeader({
      'Content-Type': 'application/json',
    }),
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    let errorData: TranscriptionErrorDetail;

    try {
      const data = await response.json();
      errorData = data.detail || data;
    } catch {
      errorData = {
        error: 'Transcription failed',
        code: 'UNKNOWN',
        recoverable: true,
        suggestion: 'Please try again',
      };
    }

    throw new TranscriptionFailedError(
      errorData.error || 'Transcription failed',
      errorData.code || 'UNKNOWN',
      errorData.recoverable ?? true,
      errorData.suggestion
    );
  }

  const result = await response.json();
  return normalizeTranscriptionResult(result);
}

export interface TranscriptionHealthStatus {
  available: boolean;
  endpoint: string;
  api_key_configured: boolean;
  error?: string;
}

export async function checkTranscriptionHealth(): Promise<TranscriptionHealthStatus> {
  try {
    const response = await fetch(`${GATEWAY_URL}/api/transcribe/health`, {
      headers: applyPreReleaseHeader(),
    });
    if (!response.ok) {
      return {
        available: false,
        endpoint: '',
        api_key_configured: false,
        error: 'Health check failed',
      };
    }
    return response.json();
  } catch (e) {
    return {
      available: false,
      endpoint: '',
      api_key_configured: false,
      error: e instanceof Error ? e.message : 'Unknown error',
    };
  }
}

async function blobToBase64(blob: Blob): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onloadend = () => {
      const base64 = reader.result as string;
      const base64Data = base64.split(',')[1];
      resolve(base64Data || '');
    };
    reader.onerror = reject;
    reader.readAsDataURL(blob);
  });
}

function normalizeTranscriptionResult(result: unknown): TranscriptionResult {
  if (Array.isArray(result)) {
    const text = result
      .map((segment) => (segment && typeof segment === 'object' ? (segment as { text?: string }).text : ''))
      .filter(Boolean)
      .join(' ')
      .trim();
    return { text };
  }

  if (result && typeof result === 'object') {
    const data = result as { text?: string; transcription?: string; language?: string; duration?: number };
    return {
      text: data.text || data.transcription || '',
      language: data.language,
      duration: data.duration,
    };
  }

  return { text: '' };
}

export type { TranscriptionOptions, TranscriptionResult, TranscriptionErrorDetail };

File: gateway/janus_gateway/routers/transcription.py
"""Transcription proxy routes."""

from typing import Any, Optional

import httpx
import structlog
from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel

from janus_gateway.config import get_settings

logger = structlog.get_logger(__name__)

router = APIRouter(prefix="/api", tags=["transcription"])


class TranscriptionRequest(BaseModel):
    """Request payload for transcription."""

    audio_b64: str
    language: Optional[str] = None


class TranscriptionResponse(BaseModel):
    """Response payload for transcription."""

    text: str
    language: Optional[str] = None
    duration: Optional[float] = None


class TranscriptionHealthResponse(BaseModel):
    """Health check response for transcription service."""

    available: bool
    endpoint: str
    api_key_configured: bool
    error: Optional[str] = None


class TranscriptionErrorDetail(BaseModel):
    """Detailed transcription error response."""

    error: str
    code: str
    recoverable: bool
    suggestion: Optional[str] = None


def create_error_response(
    error: str,
    code: str,
    recoverable: bool,
    suggestion: Optional[str] = None,
) -> dict[str, Any]:
    """Create a structured error response."""
    return {
        "error": error,
        "code": code,
        "recoverable": recoverable,
        "suggestion": suggestion,
    }


@router.get("/transcribe/health", response_model=TranscriptionHealthResponse)
async def transcription_health() -> TranscriptionHealthResponse:
    """
    Check if transcription service is available.

    Returns status of:
    - API key configuration
    - Whisper endpoint reachability
    """
    settings = get_settings()
    api_key = settings.chutes_api_key
    whisper_endpoint = settings.whisper_endpoint

    if not api_key:
        return TranscriptionHealthResponse(
            available=False,
            endpoint=whisper_endpoint,
            api_key_configured=False,
            error="CHUTES_API_KEY not configured",
        )

    # Quick ping to verify endpoint (optional, can be slow)
    try:
        async with httpx.AsyncClient(timeout=5.0) as client:
            # Just check if endpoint responds (OPTIONS or HEAD)
            response = await client.options(whisper_endpoint)
            # Any response (even 405) means endpoint is reachable
            reachable = response.status_code < 500
    except Exception as e:
        logger.warning("transcription_endpoint_unreachable", error=str(e))
        return TranscriptionHealthResponse(
            available=False,
            endpoint=whisper_endpoint,
            api_key_configured=True,
            error=f"Endpoint unreachable: {str(e)}",
        )

    return TranscriptionHealthResponse(
        available=reachable,
        endpoint=whisper_endpoint,
        api_key_configured=True,
    )


@router.post("/transcribe", response_model=TranscriptionResponse)
async def transcribe_audio(request: TranscriptionRequest) -> TranscriptionResponse:
    """Proxy transcription requests to Chutes Whisper API."""

    settings = get_settings()
    api_key = settings.chutes_api_key
    whisper_endpoint = settings.whisper_endpoint

    if not api_key:
        logger.error("transcription_not_configured")
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=create_error_response(
                error="Voice transcription is temporarily unavailable",
                code="TRANSCRIPTION_NOT_CONFIGURED",
                recoverable=False,
                suggestion="Please type your message instead",
            ),
        )

    # Validate audio data
    if not request.audio_b64:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=create_error_response(
                error="No audio data provided",
                code="MISSING_AUDIO",
                recoverable=True,
                suggestion="Please record audio before submitting",
            ),
        )

    async with httpx.AsyncClient(timeout=60.0) as client:
        try:
            logger.info("transcription_request_sent", endpoint=whisper_endpoint)

            response = await client.post(
                whisper_endpoint,
                headers={
                    "Content-Type": "application/json",
                    "Authorization": f"Bearer {api_key}",
                },
                json={
                    "language": request.language,
                    "audio_b64": request.audio_b64,
                },
            )

            logger.info("transcription_response", status_code=response.status_code)

            if response.status_code == 401:
                logger.error("transcription_invalid_api_key")
                raise HTTPException(
                    status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                    detail=create_error_response(
                        error="Voice transcription authentication failed",
                        code="INVALID_API_KEY",
                        recoverable=False,
                        suggestion="Please type your message instead",
                    ),
                )

            if response.status_code == 429:
                logger.warning("transcription_rate_limited")
                raise HTTPException(
                    status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                    detail=create_error_response(
                        error="Too many transcription requests",
                        code="RATE_LIMITED",
                        recoverable=True,
                        suggestion="Please wait a moment and try again",
                    ),
                )

            if response.status_code != 200:
                logger.error(
                    "transcription_failed",
                    status_code=response.status_code,
                    response_text=response.text,
                )
                raise HTTPException(
                    status_code=response.status_code,
                    detail=create_error_response(
                        error=f"Transcription failed: {response.text}",
                        code="UPSTREAM_ERROR",
                        recoverable=True,
                        suggestion="Please try again",
                    ),
                )

            result = response.json()
            text = result.get("text", result.get("transcription", ""))

            if not text:
                logger.warning("transcription_empty_text")

            logger.info("transcription_successful", char_count=len(text))

            return TranscriptionResponse(
                text=text,
                language=result.get("language"),
                duration=result.get("duration"),
            )

        except httpx.TimeoutException as exc:
            logger.error("transcription_timeout", error=str(exc))
            raise HTTPException(
                status_code=status.HTTP_504_GATEWAY_TIMEOUT,
                detail=create_error_response(
                    error="Transcription timed out",
                    code="TIMEOUT",
                    recoverable=True,
                    suggestion="Please try a shorter recording",
                ),
            ) from exc

        except httpx.RequestError as exc:
            logger.error("transcription_request_error", error=str(exc))
            raise HTTPException(
                status_code=status.HTTP_502_BAD_GATEWAY,
                detail=create_error_response(
                    error="Could not reach transcription service",
                    code="SERVICE_UNREACHABLE",
                    recoverable=True,
                    suggestion="Please try again in a moment",
                ),
            ) from exc
